apps/ejabberd/src/mongoose_commands.erl

```diff
@@ -0,0 +1,567 @@
+
+%% @doc make sure the command may be registered
+%% it may not if either (a) command of that name is already registered,
+%% (b) there is a command in the same category with the same action
+check_registration(Command) ->
+    Name = name(Command),
+    Cat = category(Command),
+    Act = action(Command),
+    case ets:lookup(mongoose_commands, Name) of
+        [] ->
+            CatLst = list(admin, Cat),
+            FCatLst = [C || C <- CatLst, C#mongoose_command.action == Act],
+            case FCatLst of
+                [] -> ok;
+                [C] ->
+                    baddef("There is command ~p in category ~p, action ~p", [name(C), Cat, Act])
+            end;
+        _ ->
+            ?DEBUG("This command is already defined:~n~p", [Name])
+    end.
+
+mapget(K, Map) ->
+    try maps:get(K, Map) of
+        V -> V
+    catch
+        error:{badkey, K} ->
+            th("Missing argument: ~p", [K]);
+        error:bad_key ->
+            th("Missing argument: ~p", [K])
+    end.
+
+map_to_list(Map, Args) ->
+    SpecLen = length(Args),
+    ALen = maps:size(Map),
+    if SpecLen =/= ALen ->
+        th("Invalid number of arguments: should be ~p, got ~p", [SpecLen, ALen]);
+        true -> ok
+    end,
+    [mapget(K, Map) || {K, _} <- Args].
+
+
+%% @doc Main entry point for permission control - is this command available for this user
+is_available_for(User, C) when is_binary(User) ->
+    is_available_for(jid:from_binary(User), C);
+is_available_for(admin, _C) ->
+    true;
+is_available_for(Jid, #mongoose_command{security_policy = Policies}) ->
+    apply_policies(Policies, Jid).
+
+%% @doc Check all security policies defined in the command - passes if any of them returns true
+apply_policies([], _) ->
+    false;
+apply_policies([P|Policies], Jid) ->
+    case apply_policy(P, Jid) of
+        true ->
+            true;
+        false ->
+            apply_policies(Policies, Jid)
+    end.
+
+%% @doc This is the only policy we know so far, but there will be others (like roles/acl control)
+apply_policy(user, _) ->
+    true;
+apply_policy(_, _) ->
+    false.
+
+
+locate_caller(L) ->
+    locate_caller(1, L).
+
+locate_caller(I, []) ->
+    0;
+locate_caller(I, [{caller, _}|_]) ->
+    I;
+locate_caller(I, [_|T]) ->
+    locate_caller(I + 1, T).
+
+
+check_caller(admin, _Command, _Args) ->
+    ok;
+check_caller(_Caller, #mongoose_command{caller_pos = 0}, _Args) ->
+    % no caller in args
+    ok;
+check_caller(Caller, #mongoose_command{caller_pos = CallerPos, name = Name}, Args) ->
+    % check that server and user match (we don't care about resource)
+    ACaller = lists:nth(CallerPos, Args),
+    CallerJid = jid:from_binary(Caller),
+    ACallerJid = jid:from_binary(ACaller),
+    ACal = {ACallerJid#jid.user, ACallerJid#jid.server},
+    case {CallerJid#jid.user, CallerJid#jid.server} of
+        ACal ->
+            ok;
+        _ ->
+            throw(caller_jid_mismatch)
+    end.
+
```

Comment:
I think we can save same lines by removing the unnecessary lines at the end of the file :P